store/tikv/latch/latch.go

```diff
@@ -0,0 +1,186 @@
+	if len(l.waiting) == 0 {
+		// Should we reset the lastCommitTs since next key
+		// may different from current key?
+		// l.lastCommitTs = 0
+		isEmpty = true
+	} else {
+		front = l.waiting[0]
+		isEmpty = false
+	}
+	return
+}
+
+// try append startTs into latch, return false
+// when startTs is already timeout
+func (l *Latch) append(startTs uint64) bool {
+	if startTs <= l.lastCommitTs {
+		return false
+	}
+	l.waiting = append(l.waiting, startTs)
+	return true
+}
+
+// Lock is the locks' information required for a transaction
+type Lock struct {
+	// the slot IDs of the latches(keys) that a startTs must acquire before being able to processed
+	requiredSlots []int
+	/// The number of latches that the transaction has acquired.
+	acquiredCount int
+	/// The number of latches that the waiting queue contains current transaction
+	waitedCount int
+	// Current transaction's startTs
+	startTs uint64
+}
+
+// NewLock creates a new lock
+func NewLock(startTs uint64, requiredSlots []int) Lock {
+	return Lock{
+		requiredSlots: requiredSlots,
+		acquiredCount: 0,
+		waitedCount:   0,
+		startTs:       startTs,
+	}
+}
+
+// Latches which are used for concurrency control
+// Each latch is indexed by a slit ID, hence the term latch and slot are used in interchangeable,
+// but conceptually a latch is a queue, and a slot is an index to the queue
+type Latches struct {
+	slots []Latch
+	size  int
+}
+
+// NewLatches the size will be rounded up to the power of 2
+func NewLatches(size int) Latches {
+	powerOfTwoSize := int(math.Pow(2, math.Ceil(math.Log2(float64(size)))))
+	slots := make([]Latch, powerOfTwoSize, powerOfTwoSize)
+	return Latches{
+		slots: slots,
+		size:  powerOfTwoSize,
+	}
+}
+
+// GenLock generates Lock for the transaction with startTs and keys
+func (latches *Latches) GenLock(startTs uint64, keys [][]byte) Lock {
+	hashes := make(map[int]bool)
+	for _, key := range keys {
+		hashes[latches.hash(key)] = true
+	}
+	slots := make([]int, 0, len(hashes))
+	for key := range hashes {
+		slots = append(slots, key)
+	}
+	sort.Ints(slots)
+	return NewLock(startTs, slots)
+}
+
+// return hash int for current key
+func (latches *Latches) hash(key []byte) int {
+	h := murmur3.New32()
+	_, err := h.Write(key)
+	if err != nil {
+		log.Warn("hash key %v failed with err:%+v", key, err)
+	}
+	return int(h.Sum32()) & (latches.size - 1)
+}
+
+// Acquire tries to acquire the lock for a transaction
+// It returns with timeout = true when the transaction is timeout(
+// when the lock.startTs is smaller than any key's last commitTs).
+// It returns with acquired = true when acquire success and the transaction
+// is ready to 2PC
+func (latches *Latches) Acquire(lock *Lock) (acquired, timeout bool) {
+	var newWait bool
+	for lock.acquiredCount < len(lock.requiredSlots) {
+		slotID := lock.requiredSlots[lock.acquiredCount]
+		acquired, timeout, newWait = latches.slots[slotID].acquire(lock.startTs)
+		if newWait {
+			lock.waitedCount++
```

Comment:
how about let `waitedCount` directly equal `acquiredCount+1` before return

Reply:
no, `waitedCount` may equal to `acquiredCount` when any slot is timeout